packages/ec-core/src/operator/selector/recursive_weighted.rs

```diff
@@ -1,72 +1,115 @@
-use std::marker::PhantomData;
+use rand::distr::{Bernoulli, Distribution};
 
-use rand::Rng;
-
-use super::{weighted, Selector};
+use super::Selector;
 use crate::population::Population;
 
-#[derive(Debug)]
-pub enum WeightedSelectorCreationError {
-    WeightSumOverflows(usize, usize),
+#[derive(Debug, thiserror::Error)]
+#[error("Overflow while trying to calculate the sum of the weights {0} and {1}.")]
+pub struct WeightSumOverflow(u32, u32);
+
+pub trait WeightedSelector {
+    fn weight(&self) -> u32;
+}
+
+pub struct Weighted<T> {
+    selector: T,
+    weight: u32,
 }
 
-pub trait WeightedSelector<P>: Selector<P>
+pub trait WithWeightedSelector
 where
-    P: Population,
+    Self: Sized,
 {
-    fn weight(&self) -> usize;
+    type Inner;
```

Comment:
I don't think that's a great name. It doesn't tell us anything about what kind of thing `Inner` probably is.

Reply:
maybe rename it to InnerSelector?